system/ee/ExpressionEngine/Addons/channel/libraries/channel_form/Channel_form_lib.php

```diff
@@ -2822,37 +2821,62 @@ public function initialize($reinitialize = false)
 
         $this->fetch_settings();
 
-        $this->option_fields = $this->native_option_fields;
+        if (empty($this->option_fields)) {
+
+            $cache_key = '/Fieldtype/OptionFieldtypes';
+            $this->option_fields = ee()->cache->get($cache_key);
+
+            if (empty($this->option_fields)) {
+                $fieldtypes = ee('Model')->get('Fieldtype')->all()->pluck('name');
+
+                ee()->load->library('api');
+                ee()->legacy_api->instantiate('channel_fields');
+
+                // Get the list of Fieldtypes that extend OptionFieldtype
+                $this->option_fields = array_filter($fieldtypes, function($fieldtype) {
+                    ee()->api_channel_fields->include_handler($fieldtype);
+                    $class = ucfirst($fieldtype) . '_ft';
+
+                    return is_subclass_of($class, 'OptionFieldtype');
+                });
+
+                $this->option_fields = array_values($this->option_fields);
+
+                ee()->cache->save($cache_key, $this->option_fields);
```

Comment:
@gilbert-hop I have expected EE code for similar cases, and it apears that `ee()->session->cache` is being used instead. I believe session cache does not need invalidation.

If you have strong reasons to use `ee()->cache` this would need to be invalidated when fieldtypes are removed, 'clear all caches' is clicked etc.

Reply:
I can definitely change this to be using the session cache. However, reviewing the entire logic, do we really need to cache the option fieldtypes at all? Usually, you'd only have 1 channel form tag pairs right?

Reply:
I'd say to cache is better than not to cache. It is true that usually there is just one channel form, but sometimes there are several, and they can be used multiple times during session

Reply:
@intoeetive sorry I missed the last message. The session cache is implemented now.